lib/actions/document_ai.ex

```diff
@@ -0,0 +1,98 @@
+defmodule WorkflowEngine.Actions.DocumentAi do
+  @doc """
+  Executes a request against the Azure Document AI API.
+  """
+
+  use OK.Pipe
+
+  @behaviour WorkflowEngine.Action
+
+  require Logger
+
+  @default_document_ai_api_version "2023-07-31"
+
+  def execute(state, %{"type" => "document_ai"} = step) do
+    with {:ok, model_id} <- get_property(step, "model_id"),
+         {:ok, document_url} <- get_property(step, "document_url"),
+         {:ok, api_key} <- get_property(step, "api_key"),
+         {:ok, endpoint} <- get_property(step, "endpoint"),
+         {:ok, api_version} <-
+           get_property(step, "api_version", @default_document_ai_api_version),
+         {:ok, request} <-
+           request_document_analysis(api_key, endpoint, model_id, document_url, api_version),
+         {:ok, operation_location} <- get_operation_location(request),
+         {:ok, result} <-
+           request_analyzed_results(api_key, operation_location, System.os_time(:millisecond)) do
+      # TODO: parse results and update state
```

Comment:
Result of the action should be:

```
{
  "fields": "value from document.fields",
  "confidence": "value from document.confidence",
  "analyze_result_url": operation_location
}
```

Reply:
Upon closer inspection, we do not get an overall `confidence` result. This is per document and per field in each document. In our case we only analyze one document but that can change in the future. What we can do:

```json
{
  "analyzeResult": {
    "apiVersion": "2023-07-31",
    "documents": [
      {
        "boundingRegions": [
          {
            "pageNumber": 1,
            "polygon": [0, 0, 1620, 0, 1620, 2160, 0, 2160]
          }
        ],
        "confidence": 0.516,
        "docType": "rekord-fenster-delivery-note",
        "fields": {
          "FooKey": {
            "boundingRegions": [
              {
                "pageNumber": 1,
                "polygon": [620, 1101, 817, 1104, 816, 1144, 619, 1142]
              }
            ],
            "confidence": 0.866,
            "content": "FooValue",
            "spans": [{ "length": 8, "offset": 474 }],
            "type": "string",
            "valueString": "FooValue"
          },
          "BarKey": {
            "boundingRegions": [
              {
                "pageNumber": 1,
                "polygon": [430, 1191, 562, 1190, 562, 1222, 430, 1222]
              }
            ],
            "confidence": 0.916,
            "content": "BarValue",
            "spans": [{ "length": 8, "offset": 513 }],
            "type": "string",
            "valueString": "BarValue"
          }
        }
      }
    ],
    "modelId": "rekord-fenster-delivery-note"
  },
  "createdDateTime": "2024-07-01T11:03:40Z",
  "lastUpdatedDateTime": "2024-07-01T11:03:42Z",
  "status": "succeeded"
}

```

This is after only removing unused data, no other mapping. What do you think?

Reply:
Picking the confidence from the first document sounds very acceptable to me.
The per-field confidence would be included in `fields` anyways as we said we wouldn't transform those values for now